include/LinearAlgebra/sparseLUSolver.h

```diff
@@ -103,206 +99,474 @@ SparseLUSolver<T>::SparseLUSolver(T tolerance_abs, T tolerance_rel)
-            b[i] -= L_values[idx] * b[L_col_idx[idx]];
+        // Ensure diagonal is included
+        if (marker[i] != i) {
+            marker[i] = i;
+            row_marked_indices.push_back(i);
         }
-    }
 
-    // Backward substitution (U * b = y) -> b now holds x
-    for (int i = n - 1; i >= 0; i--) {
-        T diag = 0;
-        for (int idx = U_row_ptr[i]; idx < U_row_ptr[i + 1]; idx++) {
-            int col = U_col_idx[idx];
-            if (col == i) {
-                diag = U_values[idx]; // Store diagonal value
+        // Process fill-in elements
+        while (!stk.empty()) {
+            int j = stk.back();
+            stk.pop_back();
+
+            // Process U[j] pattern
+            for (int col2 : U_pattern[j]) {
+                if (marker[col2] != i) {
+                    marker[col2] = i;
+                    row_marked_indices.push_back(col2);
+                    if (col2 < i) { // Continue for lower indices
+                        stk.push_back(col2);
+                    }
+                }
+            }
+        }
+
+        // Split into L and U patterns
+        auto& Li = L_pattern[i];
+        auto& Ui = U_pattern[i];
+        for (int col : row_marked_indices) {
+            if (col < i) {
+                Li.push_back(col);
             }
             else {
-                b[i] -= U_values[idx] * b[col];
+                Ui.push_back(col);
             }
         }
-        assert(std::abs(diag) > tolerance_abs_);
-        b[i] /= diag;
+
+        // Sort patterns for efficient access
+        std::sort(Li.begin(), Li.end());
+        std::sort(Ui.begin(), Ui.end());
     }
 }
 
+/**
+ * Numeric factorization using symbolic patterns
+ * @param A - Input matrix
+ * @param L_pattern - Symbolic pattern for L
+ * @param U_pattern - Symbolic pattern for U
+ */
 template <typename T>
-void SparseLUSolver<T>::solveInPlace(Vector<T>& b) const
+void SparseLUSolver<T>::numericFactorization(const SparseMatrixCSR<T>& A,
+                                             const std::vector<std::vector<int>>& L_pattern,
+                                             const std::vector<std::vector<int>>& U_pattern)
 {
-    assert(b.size() == static_cast<int>(L_row_ptr.size()) - 1);
-    solveInPlace(b.begin());
+    const int n = A.rows();
+
+    // Initialize storage structures
+    L_values.clear();
+    L_col_idx.clear();
+    L_row_ptr.resize(n + 1, 0);
+    U_values.clear();
+    U_col_idx.clear();
+    U_row_ptr.resize(n + 1, 0);
+    U_diag.resize(n, 0);
+
+    // Compute row pointers
+    for (int i = 0; i < n; i++) {
+        L_row_ptr[i + 1] = L_row_ptr[i] + L_pattern[i].size();
+        U_row_ptr[i + 1] = U_row_ptr[i] + U_pattern[i].size();
+    }
+
+    // Allocate memory for values and indices
+    L_values.resize(L_row_ptr[n]);
+    L_col_idx.resize(L_row_ptr[n]);
+    U_values.resize(U_row_ptr[n]);
+    U_col_idx.resize(U_row_ptr[n]);
```

Comment:
Is it useful to call `clear()` as well as `resize`?
```suggestion
    L_row_ptr.resize(n + 1, 0);
    U_row_ptr.resize(n + 1, 0);
    U_diag.resize(n, 0);

    // Compute row pointers
    for (int i = 0; i < n; i++) {
        L_row_ptr[i + 1] = L_row_ptr[i] + L_pattern[i].size();
        U_row_ptr[i + 1] = U_row_ptr[i] + U_pattern[i].size();
    }

    // Allocate memory for values and indices
    L_values.resize(L_row_ptr[n]);
    L_col_idx.resize(L_row_ptr[n]);
    U_values.resize(U_row_ptr[n]);
    U_col_idx.resize(U_row_ptr[n]);
```